File: scripts/zigux/check-phase1-reminder-companion-packet.py
#!/usr/bin/env python3
"""Guard the current Phase 1 reminder-companion checker packet against drift."""

from __future__ import annotations

import argparse
import tempfile
from pathlib import Path


HERE = Path(__file__).resolve()
DEFAULT_ROOT = HERE.parents[2] if len(HERE.parents) > 2 else HERE.parent

PHASE1_CLOSURE_REL = Path("Documentation/zigux/phase1-closure.md")
DOCS_ROOT_REL = Path("Documentation/zigux/README.md")
REVIEW_CHECKLIST_REL = Path("Documentation/zigux/review-checklist.md")
SCRIPTS_README_REL = Path("scripts/zigux/README.md")
VALIDATOR_REL = Path("scripts/zigux/validate-phase1-closure.py")
STRING_REVIEW_REL = Path("scripts/zigux/check-phase1-string-review-packet.py")
DIRECT_OWNER_REL = Path("scripts/zigux/check-phase1-direct-owner-markers.py")

REQUIRED_FILES = (
    PHASE1_CLOSURE_REL,
    DOCS_ROOT_REL,
    REVIEW_CHECKLIST_REL,
    SCRIPTS_README_REL,
    VALIDATOR_REL,
    STRING_REVIEW_REL,
    DIRECT_OWNER_REL,
)

REQUIRED_EXACT_LINES = {
    PHASE1_CLOSURE_REL: {
        "string_review_listed": "- `scripts/zigux/check-phase1-string-review-packet.py`",
        "direct_owner_listed": "- `scripts/zigux/check-phase1-direct-owner-markers.py`",
    },
    DOCS_ROOT_REL: {
        "string_review_listed": "- `scripts/zigux/check-phase1-string-review-packet.py`",
        "direct_owner_listed": "- `scripts/zigux/check-phase1-direct-owner-markers.py`",
        "direct_checks": "  * the current docs-root Phase 1 reminder packet should stay parked on the live owner-map, restored closure-side, string-review, direct-owner, and bench guards: `Documentation/zigux/phase1-closure.md` and `scripts/zigux/validate-phase1-closure.py` keep the current-master-safe closure packet explicit, `scripts/zigux/check-phase1-string-review-packet.py`, `scripts/zigux/check-phase1-direct-owner-markers.py`, and `scripts/zigux/check-phase1-bench.py` are the shipped direct checks, while `Documentation/zigux/review-checklist.md`, `zigux/tests/README.md`, and `scripts/zigux/README.md` keep the same historical-warning wording aligned around the broader missing installer, validator-first, bench-route, and replay surfaces.",
        "self_tests": "  * `python3 scripts/zigux/validate-phase1-closure.py`, `python3 scripts/zigux/check-phase1-string-review-packet.py --self-test`, `python3 scripts/zigux/check-phase1-direct-owner-markers.py --self-test`, and `python3 scripts/zigux/check-phase1-bench.py --self-test` replay the bounded current reminder checks, while the live checker routes guard the shipped Phase 1 packet without widening it back into the older closure-side or installer-companion stack.",
    },
    REVIEW_CHECKLIST_REL: {
        "packet_alignment": "  * if the change touches the shared Phase 1 host-tools closure packet, do `Documentation/zigux/phase1-closure.md`, `Documentation/zigux/phase1-host-helper-lane-sequencing.md`, `Documentation/zigux/README.md`, `Documentation/zigux/review-checklist.md`, `scripts/zigux/README.md`, `scripts/zigux/validate-phase1-closure.py`, `scripts/zigux/check-phase1-string-review-packet.py`, `scripts/zigux/check-phase1-direct-owner-markers.py`, `scripts/zigux/check-phase1-bench.py`, `scripts/zigux/check-phase1-shared-reminder-packet.py`, `zigux/tests/README.md`, `zigux/tests/fixtures/phase1_helper_manifest.json`, and `zig build phase1-host-tools-smoke --build-file zigux/tests/build.zig` still agree on the current closed-helper reminder packet, while the older validator-first, parity, bench-route, and replay names stay framed as historical packet members until current `master` materializes them again?",
    },
    SCRIPTS_README_REL: {
        "self_tests": "- `python3 scripts/zigux/validate-phase1-closure.py`, `python3 scripts/zigux/check-phase1-string-review-packet.py --self-test`, `python3 scripts/zigux/check-phase1-direct-owner-markers.py --self-test`, and `python3 scripts/zigux/check-phase1-bench.py --self-test` replay the shipped bounded Phase 1 reminder checks, and `zig build phase1-host-tools-smoke --build-file zigux/tests/build.zig` replays the shipped shared tests-root smoke route",
        "live_guards": "- `scripts/zigux/check-phase1-string-review-packet.py`, `scripts/zigux/check-phase1-direct-owner-markers.py`, `scripts/zigux/check-phase1-bench.py`, and `scripts/zigux/validate-phase1-closure.py` keep the shipped string-review, direct-owner, bench, and closure-validator packet explicit from the scripts root",
    },
    VALIDATOR_REL: {
        "reminder_packet_string": '        "scripts/zigux/check-phase1-string-review-packet.py,scripts/zigux/check-phase1-direct-owner-markers.py,"',
    },
}


def repo_root(root: str | None) -> Path:
    return Path(root).resolve() if root else DEFAULT_ROOT.resolve()


def load_text(root: Path, relative_path: Path) -> str:
    return (root / relative_path).read_text(encoding="utf-8")


def require_exact_line(text: str, label: str, line: str) -> list[str]:
    expected = line.strip()
    count = sum(1 for current_line in text.splitlines() if current_line.strip() == expected)
    return [] if count == 1 else [f"{label}:expected=1:actual={count}"]


def collect_failures(root: Path) -> list[str]:
    failures: list[str] = []
    for relative_path in REQUIRED_FILES:
        if not (root / relative_path).is_file():
            failures.append(f"missing_file:{relative_path.as_posix()}")
    if failures:
        return failures

    for relative_path, labels in REQUIRED_EXACT_LINES.items():
        text = load_text(root, relative_path)
        for label, line in labels.items():
            failures.extend(
                require_exact_line(text, f"{relative_path.as_posix()}:{label}", line)
            )
    return failures


def write_file(root: Path, relative_path: Path, text: str) -> None:
    path = root / relative_path
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(text, encoding="utf-8")


def build_sample_repo(root: Path) -> None:
    for relative_path in REQUIRED_FILES:
        labels = REQUIRED_EXACT_LINES.get(relative_path)
        if labels:
            write_file(root, relative_path, "# sample\n\n" + "\n".join(labels.values()) + "\n")
        else:
            write_file(root, relative_path, f"# sample for {relative_path.as_posix()}\n")


def mutate_line(root: Path, relative_path: Path, line: str, operation: str) -> None:
    target = root / relative_path
    text = target.read_text(encoding="utf-8")
    if operation == "remove":
        text = text.replace(line + "\n", "", 1)
    elif operation == "duplicate":
        text = text.replace(line + "\n", line + "\n" + line + "\n", 1)
    else:
        raise ValueError(f"unsupported operation: {operation}")
    target.write_text(text, encoding="utf-8")


def run_self_test() -> int:
    cases: list[tuple[str, Path | None, str | None, str]] = [("baseline", None, None, "none")]
    for relative_path, labels in REQUIRED_EXACT_LINES.items():
        for label, line in labels.items():
            cases.append((f"missing_{relative_path.name}_{label}", relative_path, line, "remove"))
            cases.append((f"duplicate_{relative_path.name}_{label}", relative_path, line, "duplicate"))
    cases.extend(
        [
            ("missing_string_review_file", STRING_REVIEW_REL, None, "missing_file"),
            ("missing_direct_owner_file", DIRECT_OWNER_REL, None, "missing_file"),
        ]
    )

    for name, relative_path, line, operation in cases:
        with tempfile.TemporaryDirectory(prefix=f"phase1-reminder-companion-{name}-") as tmpdir:
            root = Path(tmpdir)
            build_sample_repo(root)
            if relative_path is not None:
                target = root / relative_path
                if operation == "missing_file":
                    target.unlink()
                else:
                    assert line is not None
                    mutate_line(root, relative_path, line, operation)

            failures = collect_failures(root)
            if name == "baseline":
                if failures:
                    print(f"phase1-reminder-companion-self-test:{name}:unexpected={failures}")
                    return 1
            elif not failures:
                print(f"phase1-reminder-companion-self-test:{name}:expected_failure")
                return 1

    print("PHASE1_REMINDER_COMPANION_PACKET_SELF_TEST=pass")
    print(f"PHASE1_REMINDER_COMPANION_PACKET_SELF_TEST_CASE_COUNT={len(cases)}")
    return 0


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--root", help="override the repository root for validation")
    parser.add_argument("--self-test", action="store_true", help="run the built-in checker self-test")
    args = parser.parse_args()

    if args.self_test:
        return run_self_test()

    failures = collect_failures(repo_root(args.root))
    if failures:
        for failure in failures:
            print(failure)
        return 1

    print("PHASE1_REMINDER_COMPANION_PACKET=pass")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/zigux/check_phase1_reminder_companion_packet.zig
// Ported from check-phase1-reminder-companion-packet.py by port_phase1_guards.py
const std = @import("std");
const Io = std.Io;
const guard = @import("zigux_guard.zig");

pub const pass_marker = "PHASE1_REMINDER_COMPANION_PACKET_SELF_TEST=pass";

const DIRECT_OWNER_REL = "scripts\\zigux/check_phase1_direct_owner_markers.zig";

const DOCS_ROOT_REL = "Documentation/zigux/README.md";

const PHASE1_CLOSURE_REL = "Documentation/zigux/phase1-closure.md";

const REQUIRED_EXACT_LINES_ENTRIES = [_]struct { file: []const u8, marker: []const u8 }{
    .{ .file = "Documentation/zigux/phase1-closure.md", .marker = "- `scripts\\zigux/check_phase1_string_review_packet.zig`" },
    .{ .file = "Documentation/zigux/phase1-closure.md", .marker = "- `scripts\\zigux/check_phase1_direct_owner_markers.zig`" },
    .{ .file = "Documentation/zigux/README.md", .marker = "- `scripts\\zigux/check_phase1_string_review_packet.zig`" },
    .{ .file = "Documentation/zigux/README.md", .marker = "- `scripts\\zigux/check_phase1_direct_owner_markers.zig`" },
    .{ .file = "Documentation/zigux/README.md", .marker = "  * the current docs-root Phase 1 reminder packet should stay parked on the live owner-map, restored closure-side, string-review, direct-owner, and bench guards: `Documentation/zigux/phase1-closure.md` and `scripts\\zigux/validate_phase1_closure.zig` keep the current-master-safe closure packet explicit, `scripts\\zigux/check_phase1_string_review_packet.zig`, `scripts\\zigux/check_phase1_direct_owner_markers.zig`, and `scripts\\zigux/check_phase1_bench.zig` are the shipped direct checks, while `Documentation/zigux/review-checklist.md`, `zigux/tests/README.md`, and `scripts/zigux/README.md` keep the same historical-warning wording aligned around the broader missing installer, validator-first, bench-route, and replay surfaces." },
    .{ .file = "Documentation/zigux/README.md", .marker = "  * `zig run scripts/zigux/validate_phase1_closure.zig`, `zig run scripts/zigux/check_phase1_string_review_packet.zig -- --self-test`, `zig run scripts/zigux/check_phase1_direct_owner_markers.zig -- --self-test`, and `zig run scripts/zigux/check_phase1_bench.zig -- --self-test` replay the bounded current reminder checks, while the live checker routes guard the shipped Phase 1 packet without widening it back into the older closure-side or installer-companion stack." },
    .{ .file = "Documentation/zigux/review-checklist.md", .marker = "  * if the change touches the shared Phase 1 host-tools closure packet, do `Documentation/zigux/phase1-closure.md`, `Documentation/zigux/phase1-host-helper-lane-sequencing.md`, `Documentation/zigux/README.md`, `Documentation/zigux/review-checklist.md`, `scripts/zigux/README.md`, `scripts\\zigux/validate_phase1_closure.zig`, `scripts\\zigux/check_phase1_string_review_packet.zig`, `scripts\\zigux/check_phase1_direct_owner_markers.zig`, `scripts\\zigux/check_phase1_bench.zig`, `scripts\\zigux/check_phase1_shared_reminder_packet.zig`, `zigux/tests/README.md`, `zigux/tests/fixtures/phase1_helper_manifest.json`, and `zig build phase1-host-tools-smoke --build-file zigux/tests/build.zig` still agree on the current closed-helper reminder packet, while the older validator-first, parity, bench-route, and replay names stay framed as historical packet members until current `master` materializes them again?" },
    .{ .file = "scripts/zigux/README.md", .marker = "- `zig run scripts/zigux/validate_phase1_closure.zig`, `zig run scripts/zigux/check_phase1_string_review_packet.zig -- --self-test`, `zig run scripts/zigux/check_phase1_direct_owner_markers.zig -- --self-test`, and `zig run scripts/zigux/check_phase1_bench.zig -- --self-test` replay the shipped bounded Phase 1 reminder checks, and `zig build phase1-host-tools-smoke --build-file zigux/tests/build.zig` replays the shipped shared tests-root smoke route" },
    .{ .file = "scripts/zigux/README.md", .marker = "- `scripts\\zigux/check_phase1_string_review_packet.zig`, `scripts\\zigux/check_phase1_direct_owner_markers.zig`, `scripts\\zigux/check_phase1_bench.zig`, and `scripts\\zigux/validate_phase1_closure.zig` keep the shipped string-review, direct-owner, bench, and closure-validator packet explicit from the scripts root" },
    .{ .file = "scripts\\zigux/validate_phase1_closure.zig", .marker = "        \"scripts\\zigux/check_phase1_string_review_packet.zig,scripts\\zigux/check_phase1_direct_owner_markers.zig,\"" },
};

const REQUIRED_FILES = [_][]const u8{
    "Documentation/zigux/phase1-closure.md",
    "Documentation/zigux/README.md",
    "Documentation/zigux/review-checklist.md",
    "scripts/zigux/README.md",
    "scripts\\zigux/validate_phase1_closure.zig",
    "scripts\\zigux/check_phase1_string_review_packet.zig",
    "scripts\\zigux/check_phase1_direct_owner_markers.zig",
};

const REVIEW_CHECKLIST_REL = "Documentation/zigux/review-checklist.md";

const SCRIPTS_README_REL = "scripts/zigux/README.md";

const STRING_REVIEW_REL = "scripts\\zigux/check_phase1_string_review_packet.zig";

const VALIDATOR_REL = "scripts\\zigux/validate_phase1_closure.zig";

fn collectFailures(
    io: Io,
    allocator: std.mem.Allocator,
    root: []const u8,
) !std.ArrayList([]const u8) {
    var failures: std.ArrayList([]const u8) = .empty;
    errdefer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }

    {
        const relative_path = "Documentation/zigux/phase1-closure.md";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
            try failures.append(allocator, issue);
        }
    }
    {
        const relative_path = "Documentation/zigux/README.md";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
            try failures.append(allocator, issue);
        }
    }
    {
        const relative_path = "Documentation/zigux/review-checklist.md";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
            try failures.append(allocator, issue);
        }
    }
    {
        const relative_path = "scripts/zigux/README.md";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
            try failures.append(allocator, issue);
        }
    }
    {
        const relative_path = "scripts\\zigux/validate_phase1_closure.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
            try failures.append(allocator, issue);
        }
    }
    {
        const relative_path = "scripts\\zigux/check_phase1_string_review_packet.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
            try failures.append(allocator, issue);
        }
    }
    {
        const relative_path = "scripts\\zigux/check_phase1_direct_owner_markers.zig";
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{relative_path});
            try failures.append(allocator, issue);
        }
    }
    if (failures.items.len > 0) return failures;

    for (REQUIRED_EXACT_LINES_ENTRIES) |entry| {
        const full_path = try guard.joinPath(allocator, root, entry.file);
        defer allocator.free(full_path);
        const text = guard.readUtf8File(io, allocator, full_path) catch |err| switch (err) {
            guard.GuardError.IOError => {
                const issue = try std.fmt.allocPrint(allocator, "missing_file:{s}", .{entry.file});
                try failures.append(allocator, issue);
                continue;
            },
            else => return err,
        };
        defer allocator.free(text);
        const label = try std.fmt.allocPrint(allocator, "{s}:{s}", .{ entry.file, entry.marker });
        defer allocator.free(label);
        try guard.appendExactTrimmedLineIssue(allocator, &failures, text, label, entry.marker);
    }

    return failures;
}

fn buildSampleRepo(io: Io, allocator: std.mem.Allocator, root: []const u8) !void {
    for (REQUIRED_EXACT_LINES_ENTRIES) |entry| {
        const full_path = try guard.joinPath(allocator, root, entry.file);
        defer allocator.free(full_path);
        const existing = guard.readUtf8File(io, allocator, full_path) catch |err| switch (err) {
            guard.GuardError.IOError => try allocator.dupe(u8, "# sample\n\n"),
            else => return err,
        };
        defer allocator.free(existing);
        const updated = try std.fmt.allocPrint(allocator, "{s}{s}\n", .{ existing, entry.marker });
        defer allocator.free(updated);
        try guard.writeUtf8File(io, full_path, updated);
    }
    for (REQUIRED_FILES) |relative_path| {
        const full_path = try guard.joinPath(allocator, root, relative_path);
        defer allocator.free(full_path);
        if (!guard.pathExists(io, full_path)) {
            try guard.writeUtf8File(io, full_path, "# placeholder\n");
        }
    }
}

fn applyMutation(io: Io, allocator: std.mem.Allocator, root: []const u8, file: []const u8, needle: []const u8, operation: []const u8) !void {
    const full_path = try guard.joinPath(allocator, root, file);
    defer allocator.free(full_path);
    const text = try guard.readUtf8File(io, allocator, full_path);
    defer allocator.free(text);
    const updated = if (std.mem.eql(u8, operation, "remove")) blk: {
        const pattern = try std.fmt.allocPrint(allocator, "{s}\n", .{needle});
        defer allocator.free(pattern);
        const index = std.mem.indexOf(u8, text, pattern) orelse return;
        break :blk try std.fmt.allocPrint(allocator, "{s}{s}", .{ text[0..index], text[index + pattern.len ..] });
    } else if (std.mem.eql(u8, operation, "duplicate")) blk: {
        const index = std.mem.indexOf(u8, text, needle) orelse return;
        break :blk try std.fmt.allocPrint(allocator, "{s}{s}\n{s}{s}", .{ text[0..index], needle, needle, text[index + needle.len ..] });
    } else return;
    defer allocator.free(updated);
    try guard.writeUtf8File(io, full_path, updated);
}

pub fn runSelfTest(io: Io, allocator: std.mem.Allocator) !u8 {
    var tmp = try guard.TempWorkspace.init(io, allocator, "selftest");
    defer tmp.deinit();
    const root = try tmp.rootPath(allocator);
    defer allocator.free(root);
    try buildSampleRepo(io, allocator, root);
    {
        var failures = try collectFailures(io, allocator, root);
        defer {
            for (failures.items) |item| allocator.free(item);
            failures.deinit(allocator);
        }
        try guard.expectSelfTest(failures.items.len == 0);
    }
    for (REQUIRED_EXACT_LINES_ENTRIES) |entry| {
        try applyMutation(io, allocator, root, entry.file, entry.marker, "remove");
        var failures = try collectFailures(io, allocator, root);
        try guard.expectSelfTest(failures.items.len > 0);
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
        try buildSampleRepo(io, allocator, root);
        try applyMutation(io, allocator, root, entry.file, entry.marker, "duplicate");
        failures = try collectFailures(io, allocator, root);
        try guard.expectSelfTest(failures.items.len > 0);
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
        try buildSampleRepo(io, allocator, root);
    }
    try guard.printLine(io, "{s}", .{pass_marker});
    try guard.printLine(io, "PHASE1_REMINDER_COMPANION_PACKET_SELF_TEST_CASE_COUNT={d}", .{@as(usize, 21)});
    return 0;
}

pub fn main(init: std.process.Init) !void {
    const allocator = init.gpa;
    const io = init.io;
    const args = try init.minimal.args.toSlice(allocator);

    var explicit_root: ?[]const u8 = null;
    var self_test = false;
    var index: usize = 1;
    while (index < args.len) : (index += 1) {
        const arg = args[index];
        if (std.mem.eql(u8, arg, "--self-test")) {
            self_test = true;
            continue;
        }
        if (std.mem.eql(u8, arg, "--root") or std.mem.eql(u8, arg, "--repo-root")) {
            if (index + 1 >= args.len) std.process.exit(2);
            index += 1;
            explicit_root = args[index];
            continue;
        }
    }

    if (self_test) {
        std.process.exit(try runSelfTest(io, allocator));
    }

    const root = if (explicit_root) |value| value else try guard.defaultRepoRoot(allocator);
    defer if (explicit_root == null) allocator.free(root);

    var failures = try collectFailures(io, allocator, root);
    defer {
        for (failures.items) |item| allocator.free(item);
        failures.deinit(allocator);
    }

    if (failures.items.len > 0) {
        try guard.printLine(io, "PHASE1_REMINDER_COMPANION_PACKET_SELF_TEST=fail", .{});
        for (failures.items) |failure| try guard.printLine(io, "{s}", .{failure});
        std.process.exit(1);
    }

    try guard.printLine(io, "{s}", .{pass_marker});
    std.process.exit(0);
}
